content/ActiveAdmin 2FA with OneLogin.md

```diff
@@ -0,0 +1,191 @@
+
+``` ruby
+# app/controllers/admin_users/sessions_controller.rb
+
+prepend_before_action :require_no_authentication, only: [:new, :create]
+prepend_before_action :allow_params_authentication!, only: :create
+prepend_before_action :verify_signed_out_user, only: :destroy
+prepend_before_action(only: [:create, :destroy]) { request.env["devise.skip_timeout"] = true }
+```
+
+** How logout works **
+At this point you have 3 options to logout:
+  - destroy your app session
+  - destroy saml session
+  - force logout from your third party auth partner
+
+For this specic case we needed to perform the first one only, because logging someone out of OneLogin means he will be logged out of a lot of apps, and there was no need to force logout someone.
+Actually it was a pain in the knee to do this, because `devise_saml_authenticable` gem add routes with a `class_eval` approach directly to `Devise` engine, leaving you with almost no way to configure which routes you really want or not. You will be asking yourself `Why would I like to remove a route?`, well... that's becasue at this point you have 2 `admin/logout` routes in your app, and we know this is not a good practice by any mean.
+I ended up with this solution as the "cleanest":
+
+``` ruby
+# config/initializers/devise.rb
+
+# place this ouside any file definition at the bottom to ensure is the last interpreted by ruby
+ActionDispatch::Routing::Mapper.class_eval do
+  protected
+  def devise_saml_authenticatable(mapping, controllers)
+    if ::Devise.saml_route_helper_prefix
+      prefix = ::Devise.saml_route_helper_prefix
+      resource :session, only: [], controller: controllers[:saml_sessions], path: '' do
+        get :new, path: 'saml/sign_in', as: "new_#{prefix}"
+        post :create, path: 'saml/auth', as: prefix
+        get :metadata, path: 'saml/metadata'
+        match :idp_sign_out, path: 'saml/idp_sign_out', as: "idp_destroy_#{prefix}", via: [:get, :post]
+      end
+    else
+      resource :session, only: [], controller: controllers[:saml_sessions], path: '' do
+        get :new, path: 'saml/sign_in', as: 'new'
+        post :create, path: 'saml/auth'
+        get :metadata, path: 'saml/metadata'
+        match :idp_sign_out, path: 'saml/idp_sign_out', via: [:get, :post]
+      end
+    end
+  end
+end
+```
+
+
+
+##
+**3) Conclusions**
+
+A project that was estimated to last over a month (or so), end ups tacking only 2 weeks because we found a great gem that solves most of our problems (besides the gem implementation was not the best).
+Besides solving this, work does not end here, I'm planning on contribute back to this awesome gem to improve its quality and support missing points we speak of. And my goal with this post besides showing how to solve a security problem, is to encourage others to contribute back when you see you can do it, without contributions like this, this gem wouldn't exist either this post.
+
+
+<!-- Docs to Markdown version 1.0β17 -->
```

Comment:
Lo que?

Reply:
No idea if this is needed, just c&p from other blog 👀 . Should I remove it then?

Reply:
Yeah, that shouldn't be necessary.